File: ReadersWriterLock/ReaderWriterLock.cs
namespace ReadersWriterLock
{
    public class ReaderWriterLock
    {
        private readonly object readerMutex = new object();
        private int currentReaderCount;

        private readonly HashSet<object> globalMutex = new HashSet<object>();
        private readonly ReadLock readerLock;
        private readonly WriteLock writerLock;

        public ReaderWriterLock()
        {
            currentReaderCount = 0;
            readerLock = new ReadLock(readerMutex, globalMutex, this);
            writerLock = new WriteLock(globalMutex);
        }

        public ILock readLock()
        {
            return readerLock;
        }

        public ILock writeLock()
        {
            return writerLock;
        }

        private bool DoesWriterOwnThisLock()
        {
            return globalMutex.Contains(writerLock);
        }

        private bool IsLockFree()
        {
            return globalMutex.Count == 0;
        }

        private class ReadLock : ILock
        {
            private readonly object readerMutex;
            private readonly HashSet<object> globalMutex;
            private readonly ReaderWriterLock parent;

            public ReadLock(object readerMutex, HashSet<object> globalMutex, ReaderWriterLock parent)
            {
                this.readerMutex = readerMutex;
                this.globalMutex = globalMutex;
                this.parent = parent;
            }

            public void Lock()
            {
                lock (readerMutex)
                {
                    parent.currentReaderCount++;
                    if (parent.currentReaderCount == 1)
                    {
                        Monitor.Enter(globalMutex);
                    }
                }
            }

            public void Unlock()
            {
                lock (readerMutex)
                {
                    parent.currentReaderCount--;

                    if (parent.currentReaderCount == 0)
                    {
                        lock (globalMutex)
                        {
                            globalMutex.Remove(this);
                            Monitor.PulseAll(globalMutex);
                        }
                    }
                }
            }
        }

        private class WriteLock : ILock
        {
            private readonly HashSet<object> globalMutex;

            public WriteLock(HashSet<object> globalMutex)
            {
                this.globalMutex = globalMutex;
            }

            public void Lock()
            {
                lock (globalMutex)
                {
                    while (globalMutex.Count != 0)
                    {
                        Monitor.Wait(globalMutex);
                    }
                    globalMutex.Add(this);
                }
            }

            public void Unlock()
            {
                lock (globalMutex)
                {
                    globalMutex.Remove(this);
                    Monitor.PulseAll(globalMutex);
                }
            }
        }
    }
}

File: Java/ReadersWriterLockJava/src/ReaderWriterLock.java
import java.util.HashSet;

class ReaderWriterLock {
    private int currentReaderCount;
    private final HashSet<Object> globalMutex = new HashSet<>();
    private final ReadLock readerLock;
    private final WriteLock writerLock;

    public ReaderWriterLock() {
        currentReaderCount = 0;
        Object readerMutex = new Object();
        readerLock = new ReadLock(readerMutex, globalMutex, this);
        writerLock = new WriteLock(globalMutex);
    }

    public ILock readLock() {
        return readerLock;
    }

    public ILock writeLock() {
        return writerLock;
    }

    private boolean doesWriterOwnThisLock() {
        return globalMutex.contains(writerLock);
    }

    private boolean isLockFree() {
        return globalMutex.isEmpty();
    }

    private record ReadLock(Object readerMutex, HashSet<Object> globalMutex, ReaderWriterLock parent) implements ILock {

        @Override
            public void lock() {
                synchronized (readerMutex) {
                    parent.currentReaderCount++;
                    if (parent.currentReaderCount == 1) {
                        synchronized (globalMutex) {
                            globalMutex.add(this);
                        }
                    }
                }
            }

            @Override
            public void unlock() {
                synchronized (readerMutex) {
                    parent.currentReaderCount--;

                    if (parent.currentReaderCount == 0) {
                        synchronized (globalMutex) {
                            globalMutex.remove(this);
                            globalMutex.notifyAll();
                        }
                    }
                }
            }
        }

    private record WriteLock(HashSet<Object> globalMutex) implements ILock {

        @Override
            public void lock() {
                synchronized (globalMutex) {
                    while (!globalMutex.isEmpty()) {
                        try {
                            globalMutex.wait();
                        } catch (InterruptedException e) {
                            Thread.currentThread().interrupt();
                        }
                    }
                    globalMutex.add(this);
                }
            }

            @Override
            public void unlock() {
                synchronized (globalMutex) {
                    globalMutex.remove(this);
                    globalMutex.notifyAll();
                }
            }
        }
}